File: jlox/src/main/java/com/tre3p/REPL.java
package com.tre3p;

import com.tre3p.interpreter.Interpreter;
import com.tre3p.parser.Parser;
import com.tre3p.scanner.TokenScanner;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Paths;

public class REPL {
    private static final String PROMPT = "> ";
    private static final String USAGE_PROMPT = "Usage: jlox [script]";
    private static final Interpreter interpreter = new Interpreter();

    public static void main(String[] args) throws IOException {
        if (args.length > 1) {
            System.out.println(USAGE_PROMPT);
            System.exit(64);
        } else if (args.length == 1) {
            runFile(args[0]);
        } else {
            runPrompt();
        }
    }

    private static void run(String inputCode) {
        var tokens = new TokenScanner(inputCode).scanTokens();
        var statements = new Parser(tokens).parse();
        interpreter.interpret(statements);
    }

    private static void runFile(String filePath) throws IOException {
        run(new String(Files.readAllBytes(Paths.get(filePath))));
        if (Lox.hadError) System.exit(65);
        if (Lox.hadRuntimeError) System.exit(70);
    }

    private static void runPrompt() throws IOException {
        var br = new BufferedReader(new InputStreamReader(System.in));

        while (true) {
            System.out.println(PROMPT);

            var line = br.readLine();
            if (line == null) return;
            run(line);

            Lox.hadError = false;
        }
    }
}

File: lox-kt/src/main/kotlin/REPL.kt
import scanner.Scanner
import java.io.BufferedReader
import java.io.InputStreamReader
import java.nio.file.Files
import java.nio.file.Paths
import kotlin.system.exitProcess

private const val USAGE_PROMPT = "Usage: ktlox [script]"
private const val PROMPT = "> "

fun main(args: Array<String>) {
    if (args.size > 1) {
        println(USAGE_PROMPT)
        exitProcess(64)
    } else if (args.size == 1) {
        runFile(args.first())
    } else {
        runPrompt()
    }
}

fun run(code: String) =
    Scanner(code)
        .scanTokens()
        .forEach { println(it) }

fun runFile(filePath: String): Unit =
    run(String(Files.readAllBytes(Paths.get(filePath))))

private tailrec fun runPrompt(
    inputReader: InputStreamReader = InputStreamReader(System.`in`),
    bufferedReader: BufferedReader = BufferedReader(inputReader)
) {
    println(PROMPT)

    bufferedReader.readLine().let { line ->
        if (line == null) return
        run(line)
    }

    runPrompt(inputReader, bufferedReader)
}